connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java

```diff
@@ -866,4 +867,19 @@ public List<ConfigKeyInfo> connectorPluginConfig(String pluginName) {
         }
     }
 
+    @Override
+    public void connectorOffsets(String connName, Callback<ConnectorOffsets> cb) {
+        log.debug("Submitting offset fetch request for connector: {}", connName);
+        connectorExecutor.submit(() -> {
```

Comment:
Now that the `Worker` API is callback based, can we move any dispatching of tasks to executors (or different threads in general) to there and invoke `Worker::connectorOffsets` directly in the calling thread for this method?

Reply:
Thanks for the detailed reply!

> With regards to using up threads threads on the Worker class's executor, you've correctly noted that this one is responsible for running Connector and Task instances. However, that executor doesn't have a bound on the number of threads it allocates, so we don't have to worry about tying it up and preventing it from being used for other purposes. It's possible that an excessive number of offset reset requests might cause problems for the whole JVM process by making too many calls to the executor and spinning up too many new threads, but that's equally possible if we use the AbstractHerder class's connectorExecutor, since that too uses an unbounded number of threads

Yeah, that's right but my concern wasn't about resource contention between connector / task threads and offset read requests but more so about logical separation of responsibilities. I considered adding a new unbounded cached thread pool `requestExecutor` to the `Worker` to handle offset read requests but it doesn't really offer much other than a separate name and comes with the additional burden of having to manage the new executor's lifecycle (closing cleanly on worker termination etc.). Also, I just noticed that the existing worker executor is not exclusively used to run connector and task threads, but it's also passed to worker source tasks to close producers on a separate thread during task cancellation. So, I've elected to just simply re-use the existing `executor` (which is anyway generically named) in the `Worker` for servicing the async offset get request as well (and we now no longer make use of the executor in the `AbstractHerder`). Let me know what you think!